source/NVDAObjects/UIA/winConsoleUIA.py

```diff
@@ -13,43 +13,116 @@
+				# We must pull back the end by one character otherwise when we collapse to end,
+				# a console bug results in a textRange covering the entire console buffer!
+				# Strangely the *very* last character is a special blank point
+				# so we never seem to miss a real character.
+				UIATextInfo.move(tempInfo, textInfos.UNIT_CHARACTER, -1, endPoint="end")
+				tempInfo.setEndPoint(tempInfo, "startToEnd")
+				_rangeObj = tempInfo._rangeObj
 		elif position is textInfos.POSITION_ALL:
-			first = self.__class__(obj, textInfos.POSITION_FIRST)
-			last = self.__class__(obj, textInfos.POSITION_LAST)
-			first.setEndPoint(last, "endToEnd")
-			_rangeObj = first._rangeObj
+			if visiRange:
+				_rangeObj = visiRange
+			else:
+				first = self.__class__(obj, textInfos.POSITION_FIRST)
+				last = self.__class__(obj, textInfos.POSITION_LAST)
+				first.setEndPoint(last, "endToEnd")
+				_rangeObj = first._rangeObj
 		super(consoleUIATextInfo, self).__init__(obj, position, _rangeObj)
+		if visiRange and position is textInfos.POSITION_FIRST:
+			self.collapse(end=False)
+		elif visiRange and position is textInfos.POSITION_LAST:
+			self.move(textInfos.UNIT_CHARACTER, -1, endPoint="end")
+			self.collapse(end=True)
+
+	def move(self, unit, direction, endPoint=None):
+		oldInfo = None
+		if self.basePosition != textInfos.POSITION_CARET:
+			# Ensure we haven't gone beyond the visible text.
+			# UIA adds thousands of blank lines to the end of the console.
+			boundingInfo = self.obj.makeTextInfo(textInfos.POSITION_ALL)
+			oldInfo = self.copy()
+		res = self._move(unit, direction, endPoint)
+		# Console textRanges have access to the entire console buffer.
+		# However, we want to limit ourselves to onscreen text.
+		# Therefore, if the textInfo was originally visible,
+		# but we are now above or below the visible range,
+		# Restore the original textRange and pretend the move didn't work.
+		if oldInfo:
+			try:
+				if (
+					(
+						self.compareEndPoints(boundingInfo, "startToStart") < 0
+						or self.compareEndPoints(boundingInfo, "startToEnd") >= 0
+					)
+					and not (
+						oldInfo.compareEndPoints(boundingInfo, "startToStart") < 0
+						or oldInfo.compareEndPoints(boundingInfo, "startToEnd") >= 0
+					)
+				):
+					self._rangeObj = oldInfo._rangeObj
+					return 0
+			except (COMError, RuntimeError):
+				pass
+		return res
+
+	def _move(self, unit, direction, endPoint=None):
+		"Perform a move without respect to bounding."
+		return super(consoleUIATextInfo, self).move(unit, direction, endPoint)
+
+	def __ne__(self, other):
+		"""Support more accurate caret move detection."""
+		return not self == other
+
+	def _get_text(self):
+		# #10036: return a space if the text range is empty.
+		# Consoles don't actually store spaces, the character is merely left blank.
+		res = super(consoleUIATextInfo, self)._get_text()
+		if not res:
+			return ' '
+		else:
+			return res
+
 
+class consoleUIATextInfoEndInclusive(consoleUIATextInfo):
```

Comment:
I have the feeling that this does more than being end inclusive, may be something like this

```suggestion
class consoleUIATextInfoPre2004(consoleUIATextInfo):
```